nixos/lib/test-driver/lib/driver.py

```diff
@@ -0,0 +1,175 @@
+        vm_scripts,
+        keep_vm_state=False,
+        configure_python_repl=id,
+        machine_class=Machine,
+        vlan_class=VLan,
+    ):
+        """
+        Args:
+            - configure_python_repl: a function to configure ptpython.repl
+        """
+        self.log = logger.log_machinestate
+        self.configure_python_repl = configure_python_repl
+
+        tmp_dir = os.environ.get("TMPDIR", tempfile.gettempdir())
+        os.makedirs(tmp_dir, mode=0o700, exist_ok=True)
+
+        self.vlans = [
+            self.log(f"starting VDE switch for network {nr}") and
+            vlan_class(nr, tmp_dir) for nr in
+            list(dict.fromkeys(os.environ.get("VLANS", "").split()))
+        ]
+
+        def cmd(scripts):
+            for s in scripts:
+                yield NixStartScript(s)
+
+        self.machines = [
+            self.log(f"creating VM '{cmd.machine_name}'") and
+            machine_class(
+                start_command=cmd,
+                keep_vm_state=keep_vm_state,
+                name=cmd.machine_name,
+                log_serial=logger.log_serial,
+                log_machinestate=logger.log_machinestate,
+                tmp_dir=tmp_dir,
+            ) for cmd in cmd(vm_scripts)
+        ]
+
+        @atexit.register
+        def clean_up() -> None:
+            with self.log.nested("cleaning up"):
+                for machine in self.machines:
+                    if not machine.release():
+                        continue
+                    self.log(f"killed {machine.name} (pid {machine.pid})")
+                for vlan in self.vlans:
+                    vlan.release()
+                    self.log(f"killed {machine.name} (pid {machine.pid})")
+            self.log.release()
+
+    @contextmanager
+    def subtest(self, name: str) -> Iterator[None]:
+        with self.log.nested(name):
+            try:
+                yield
+                return True
+            except Exception as e:
+                self.log.log(f'Test "{name}" failed with error: "{e}"')
+                raise e
+
+        return False
+
+    def export_symbols(self):
+        global machines
+        machines = self.machines
+        machine_eval = [
+            "global {0}; {0} = machines[{1}]".format(m.name, idx) for idx, m in enumerate(machines)
+        ]
+
+        exec("\n".join(machine_eval))
+
+        global start_all, test_script
+        start_all = self.start_all
+        test_script = self.test_script
+
+        global subtest
+
+        def subtest(name):
+            self.subtest(name)
+
+    def test_script(self) -> None:
+        with self.log.nested("running the VM test script"):
+            exec(os.environ["testScript"])
+
+    def run_tests(self) -> None:
+        tests = os.environ.get("tests", None)
+        if tests is not None:
+            with self.log.nested("running the VM test script"):
+                try:
+                    exec(tests, globals())
+                except Exception:
```

Comment:
Any particular reason we shouldn't just let this exception propagage? (genuinely asking)

Reply:
done, not sure though. We can always revert if somebody comes up with a good argument. :smile: